### PR TITLE
Refactor FXIOS-5843 [v113] Address assertions on main thread

### PR DESCRIPTION
**Pull request by nishant2718**
# [FXIOS-5843](https://mozilla-hub.atlassian.net/browse/FXIOS-5843)

TLDR: Address assert on mains by replacing them with `ensureMainThread`. 

Note that the assertMain in `RustFirefoxAccounts` was left out because that case interacts with deferred.

UPDATE Mar 3, 2023
TabManagerStore's restoreTabs method exits with a `return nil` if the operation isn't run on main thread. I believe it must run on main because TabManager runs the restore to find the tab to select and show a user. However, when given back a `nil`, TabManager's restore tabs method can handle that case and find a way out of that error state. 

Rather than keeping the assertion to allow this error state to continue in release builds, I prefer the above.

### CODE REVIEW

**Review by OrlaM — APPROVED** [review]

**Review by adudenamedruby — APPROVED** [review]

**Comment by nishant2718** [comment]
Let this sit in beta for a while to see if introduces some strange behavior.

**Comment by mte-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Project coverage: 30.19%</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>Edited 10 files</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>Created 0 files</td>
    </tr>
  </tbody>
</table>

## Client.app: Coverage: 27.11
| File | Coverage ||
| --- | --- | --- |
SavedTab+ConfigureExtension.swift | 68.24% | ✅
RemoteTabsPanel.swift | 68.75% | ✅
LoginListViewModel.swift | 17.57% | ⚠️
SearchViewController.swift | 0.0% | ⚠️
TopTabsViewController.swift | 0.0% | ⚠️
TabManagerStore.swift | 79.53% | ✅
TabLocationView.swift | 41.59% | ⚠️
BrowserViewController.swift | 20.5% | ⚠️
DispatchQueueHelper.swift | 50.0% | ⚠️


## CredentialProvider.appex: Coverage: 18.77
| File | Coverage ||
| --- | --- | --- |
DispatchQueueHelper.swift | 50.0% | ⚠️

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against aafd5913032b886b4dbee47b43dca56dbe616947
</p>

**Comment by nishant2718** [comment]
Even after weakifying all self references in those closures, it seems like this fails on an unrelated test, `JumpBackInDataAdaptorTests.testEmptyData_tabTrayGroupsDisabled()`. It looks like this test is flaky on `main`.

**Review by adudenamedruby — APPROVED** [review]

**Review by OrlaM — APPROVED** [review]
